File: src/crwallm/api/routers/recipes.py
"""Recipe endpoints.

**Writing** a recipe is not here. Making one means fetching a sample page,
detecting its structure, asking a model to name the columns and scoring
several candidates - minutes, with a model in the loop, which is what the job
queue exists for (docs/07_RECIPE_ARCHITECTURE.md).

**Testing** one is. It fetches a single page and scores what the recipe
already says, with no model and no queue: a couple of seconds, which is an
HTTP request. That distinction is the whole reason ``test`` and ``activate``
are here while ``adapt`` is not - and it matters, because a screen that can
list recipes but never prove one turns "active" into a word nobody can check.

The UI needs the list so it can offer recipes when submitting a crawl, and it
needs the quality numbers so a stale recipe is visible as stale rather than
just "active".
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Body, Depends, HTTPException, status

from crwallm.api.deps import settings_dep, token_dep
from crwallm.api.schemas import RecipeDetail, RecipeSummary
from crwallm.config import Settings
from crwallm.schemas.recipe import Recipe
from crwallm.services.recipe import RecipeFileError, RecipeStore

router = APIRouter(prefix="/api/recipes", tags=["recipes"])

Config = Annotated[Settings, Depends(settings_dep)]


def _summary(recipe: Recipe) -> RecipeSummary:
    return RecipeSummary(
        name=recipe.name,
        version=recipe.version,
        status=recipe.status.value,
        source=recipe.source,
        source_url=recipe.source_url,
        allowed_domains=list(recipe.allowed_domains),
        container=recipe.container,
        field_names=[f.name for f in recipe.fields],
        record_count=recipe.quality.record_count,
        mean_fill=recipe.quality.mean_fill,
        measured_at=recipe.quality.measured_at,
    )


@router.get("", response_model=list[RecipeSummary])
def list_recipes(settings: Config) -> list[RecipeSummary]:
    """Every recipe on disk.

    Files that fail to load are skipped rather than raising: one recipe with a
    typo should not blank the list of the twenty that work. The CLI's
    ``recipe list`` is where the parse errors are shown.
    """
    loaded, _errors = RecipeStore(settings.recipes_dir).load_all()
    return [_summary(r) for r in loaded]


@router.get("/{name}", response_model=RecipeDetail)
def get_recipe(name: str, settings: Config) -> RecipeDetail:
    try:
        recipe = RecipeStore(settings.recipes_dir).load(name)
    except RecipeFileError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc

    return RecipeDetail(
        **_summary(recipe).model_dump(),
        fields=[f.model_dump(mode="json", exclude_none=True) for f in recipe.fields],
        fingerprint=recipe.fingerprint,
        notes=recipe.notes,
    )


# --------------------------------------------------------------- proving one


def _quality(result: object) -> dict[str, object]:
    """The measurement, flattened for a screen.

    Every number the activation gate looks at, so a refusal can be read
    against the same figures the gate used rather than taken on faith.
    """
    from typing import Any, cast

    r = cast(Any, result)
    return {
        "passes": r.passes,
        "summary": r.summary,
        "container_matched": r.quality.container_matched,
        "record_count": r.quality.record_count,
        "mean_fill": r.quality.mean_fill,
        "consistency": r.quality.consistency,
        "score": r.quality.score,
        "fill_rates": dict(sorted(r.quality.fill_rates.items())),
        "sample_url": r.quality.sample_url,
        "filtered_out": r.filtered_out,
        "filter_reasons": dict(r.filter_reasons),
        "records": [dict(row) for row in r.records[:5]],
    }


async def _measure(recipe: Recipe, url: str | None) -> object:
    from crwallm.crawler.contracts import FetchFailure, FetchRequest
    from crwallm.crawler.fetching.http import SafeHttpFetcher
    from crwallm.policy.local import build_guard
    from crwallm.policy.url import normalize
    from crwallm.schemas.types import FetchMode
    from crwallm.services.recipe import measure

    target = url or recipe.source_url
    fetcher = SafeHttpFetcher(build_guard())
    try:
        outcome = await fetcher.fetch(
            FetchRequest(
                url=normalize(target),
                depth=0,
                mode=FetchMode.HTTP,
                timeout_s=20.0,
                byte_limit=8_000_000,
            )
        )
    finally:
        await fetcher.aclose()

    if isinstance(outcome, FetchFailure):
        from crwallm.services.quick import fetch_error

        raise HTTPException(status.HTTP_502_BAD_GATEWAY, detail=fetch_error(outcome))
    return measure(recipe, outcome)


@router.post("/{name}/test", dependencies=[Depends(token_dep)])
async def test_recipe(
    name: str,
    settings: Config,
    url: Annotated[str | None, Body(embed=True)] = None,
) -> dict[str, object]:
    """Run the recipe against one page and report what it got.

    Changes nothing on disk. This is the button to press before believing a
    recipe, and after a site has been restyled.
    """
    try:
        recipe = RecipeStore(settings.recipes_dir).load(name)
    except RecipeFileError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc

    return _quality(await _measure(recipe, url))


@router.post("/{name}/activate", dependencies=[Depends(token_dep)])
async def activate_recipe(
    name: str,
    settings: Config,
    url: Annotated[str | None, Body(embed=True)] = None,
) -> dict[str, object]:
    """Re-measure, then promote to ``active`` if the evidence supports it.

    Re-measured rather than trusting the file: ``active`` is a claim that this
    works now, and the numbers already written there were true whenever they
    were taken.
    """
    from crwallm.services.recipe import activate

    store = RecipeStore(settings.recipes_dir)
    try:
        recipe = store.load(name)
    except RecipeFileError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc

    result = await _measure(recipe, url)
    report = _quality(result)
    try:
        promoted = activate(result)  # type: ignore[arg-type]
    except ValueError as exc:
        # 409, not 422: the request was well formed and the recipe is real -
        # it simply did not earn the promotion, and the numbers above say why.
        raise HTTPException(
            status.HTTP_409_CONFLICT, detail=f"활성화할 수 없습니다: {exc}"
        ) from None

    store.save(promoted)
    return {**report, "status": promoted.status.value, "version": promoted.version}

File: src/crwallm/ui/recipes.js
/*
 * 레시피: what this tool has learned about which sites.
 *
 * A recipe is the difference between understanding a site once and
 * understanding it every time. Making one can cost a model call; running it
 * never does, and that is the whole economics of the tool.
 *
 * Recipes are YAML files on disk, so this view needs no database - the only
 * screen besides 모으기 that works with Docker switched off.
 *
 * **Making** one is still `crwallm recipe adapt`: it asks a model to name the
 * columns, scores several candidates and takes minutes. **Proving** one is
 * here, because that is a single fetch and a score with no model in it - and
 * a screen that can list recipes but never check one turns 동작 확인됨 into a
 * word nobody can verify. 활성화 re-measures first for the same reason: the
 * numbers already in the file were true whenever they were taken.
 */

const recipesState = { open: null, busy: false };

/* The three the schema actually defines. Guessed names were shown as-is until
 * a real list came back reading "candidate", which tells a Korean-speaking
 * user nothing at all. */
const RECIPE_STATUS = {
  candidate: "확인 전",
  active: "동작 확인됨",
  deprecated: "물러남",
};

async function loadRecipes() {
  const empty = $("recipes-empty");
  try {
    const recipes = await rest.get("/api/recipes");
    renderRecipes(recipes);
    empty.hidden = recipes.length > 0;
    if (!recipes.length) {
      empty.textContent =
        "아직 레시피가 없습니다. 터미널에서 crwallm recipe adapt <이름> --url <주소> 로 만듭니다.";
    }
  } catch (err) {
    empty.hidden = false;
    empty.textContent = `레시피를 읽지 못했습니다: ${err.message || err}`;
  }
}

function renderRecipes(recipes) {
  const head = $("recipes-table").querySelector("thead");
  const body = $("recipes-table").querySelector("tbody");
  head.replaceChildren();
  body.replaceChildren();

  const headRow = document.createElement("tr");
  for (const name of ["이름", "상태", "어디서", "뽑은 건수", "채움", ""]) {
    const th = document.createElement("th");
    th.textContent = name;
    headRow.append(th);
  }
  head.append(headRow);

  for (const recipe of recipes) {
    const tr = document.createElement("tr");

    const name = document.createElement("td");
    name.textContent = recipe.name;
    tr.append(name);

    const status = document.createElement("td");
    const chip = document.createElement("span");
    chip.className = `chip chip-${recipe.status}`;
    chip.textContent = RECIPE_STATUS[recipe.status] || recipe.status;
    status.append(chip);
    tr.append(status);

    const source = document.createElement("td");
    source.textContent = (recipe.allowed_domains || []).join(", ") || recipe.source;
    tr.append(source);

    const count = document.createElement("td");
    // 0 and "never measured" are different answers and the second one is the
    // one that explains why activate refused.
    count.textContent = recipe.measured_at == null ? "측정 안 함" : String(recipe.record_count);
    tr.append(count);

    const fill = document.createElement("td");
    // mean_fill is a computed average and comes back 0.0 when nothing was ever
    // measured, so it cannot say for itself whether it means anything. Only
    // measured_at can, and both columns have to agree - "측정 안 함" beside
    // "0%" reads as a recipe that found nothing rather than one never run.
    fill.textContent =
      recipe.measured_at == null ? "" : `${Math.round((recipe.mean_fill || 0) * 100)}%`;
    tr.append(fill);

    const actions = document.createElement("td");
    const open = document.createElement("button");
    open.textContent = "열기";
    open.addEventListener("click", () => openRecipe(recipe.name));
    actions.append(open);
    tr.append(actions);

    body.append(tr);
  }
}

async function openRecipe(name) {
  recipesState.open = name;
  $("recipe-detail").hidden = false;
  $("recipe-detail-title").textContent = name;
  $("recipe-report").hidden = true;
  $("recipe-detail-error").hidden = true;

  try {
    const recipe = await rest.get(`/api/recipes/${encodeURIComponent(name)}`);
    const bits = [
      `source: ${recipe.source}`,
      recipe.container ? `container: ${recipe.container}` : "",
      recipe.source_url ? `기준 페이지: ${recipe.source_url}` : "",
    ].filter(Boolean);
    $("recipe-detail-meta").textContent = bits.join("  ·  ");
    renderRecipeFields(recipe.fields || []);
  } catch (err) {
    $("recipe-detail-meta").textContent = String(err.message || err);
    $("recipe-fields").querySelector("tbody").replaceChildren();
  }
  $("recipe-detail").scrollIntoView({ behavior: "smooth", block: "nearest" });
}

function renderRecipeFields(fields) {
  const head = $("recipe-fields").querySelector("thead");
  const body = $("recipe-fields").querySelector("tbody");
  head.replaceChildren();
  body.replaceChildren();

  const headRow = document.createElement("tr");
  for (const name of ["이름", "어디서", "종류", "다듬기"]) {
    const th = document.createElement("th");
    th.textContent = name;
    headRow.append(th);
  }
  head.append(headRow);

  for (const field of fields) {
    const tr = document.createElement("tr");
    for (const value of [
      field.name,
      field.selector,
      field.type || "text",
      (field.transform || []).join(", "),
    ]) {
      const td = document.createElement("td");
      td.textContent = value || "";
      td.title = td.textContent;
      tr.append(td);
    }
    body.append(tr);
  }
}

$("recipes-refresh").addEventListener("click", loadRecipes);
$("recipe-close").addEventListener("click", () => {
  recipesState.open = null;
  $("recipe-detail").hidden = true;
});


/* ------------------------------------------------------- 확인하고 활성화 */

/** Run the recipe against its sample page, or against that and then promote. */
async function proveRecipe(promote) {
  const name = recipesState.open;
  if (!name || recipesState.busy) return;

  recipesState.busy = true;
  $("recipe-detail-error").hidden = true;
  $("recipe-test").disabled = true;
  $("recipe-activate").disabled = true;
  busy(true, promote ? "다시 재보고 활성화하는 중…" : "지금도 되는지 재보는 중…");

  try {
    const verb = promote ? "activate" : "test";
    const report = await rest.post(`/api/recipes/${encodeURIComponent(name)}/${verb}`, {});
    renderReport(report, promote);
    if (promote) {
      toast(`${name}을(를) 활성화했습니다.`);
      loadRecipes();
    }
  } catch (err) {
    // A refusal is the interesting answer here, not an accident: it means the
    // recipe did not earn the promotion, and the report above already says
    // which number fell short.
    $("recipe-detail-error").hidden = false;
    $("recipe-detail-error").textContent = String(err.message || err);
  } finally {
    recipesState.busy = false;
    $("recipe-test").disabled = false;
    $("recipe-activate").disabled = false;
    busy(false);
  }
}

function renderReport(report, promoted) {
  const box = $("recipe-report");
  box.hidden = false;
  box.className = `report ${report.passes ? "ok" : "bad"}`;
  box.replaceChildren();

  const headline = document.createElement("strong");
  headline.textContent = report.passes
    ? promoted
      ? "활성화했습니다."
      : "지금도 잘 됩니다."
    : "기준에 못 미칩니다.";
  box.append(headline);

  const list = document.createElement("dl");
  const rows = [
    ["뽑은 건수", String(report.record_count)],
    ["채움", `${Math.round((report.mean_fill || 0) * 100)}%`],
    ["일관성", `${Math.round((report.consistency || 0) * 100)}%`],
    ["점수", String(report.score)],
    ["기준 페이지", report.sample_url || ""],
  ];
  if (!report.container_matched) {
    rows.unshift(["반복 단위", "찾지 못했습니다 — 사이트가 바뀌었을 수 있습니다"]);
  }
  if (report.filtered_out) {
    rows.push(["필터로 걸러짐", `${report.filtered_out}건`]);
  }
  // Per-field fill is the number that says *which* selector went stale, which
  // is the whole reason to press the button after a site is restyled.
  for (const [field, rate] of Object.entries(report.fill_rates || {})) {
    rows.push([`　${field}`, `${Math.round(rate * 100)}%`]);
  }

  for (const [label, value] of rows) {
    const dt = document.createElement("dt");
    dt.textContent = label;
    const dd = document.createElement("dd");
    dd.textContent = value;
    list.append(dt, dd);
  }
  box.append(list);
}

$("recipe-test").addEventListener("click", () => proveRecipe(false));
$("recipe-activate").addEventListener("click", () => proveRecipe(true));
